File: 63/main.rs
struct Solution{}

impl Solution {
    pub fn unique_paths_with_obstacles(obstacle_grid: Vec<Vec<i32>>) -> i32 {
        let n = obstacle_grid.len();
        let m = obstacle_grid[0].len();
        let mut f = vec![0; m];
        f[0] = if obstacle_grid[0][0] == 0 { 1 } else { 0 };
        for i in 0..n {
            for j in 0..m {
                if obstacle_grid[i][j] == 1 {
                    f[j] = 0;
                } else if j > 0 && obstacle_grid[i][j-1] == 0 {
                    f[j] += f[j-1];
                }
            }
        }
        f[m-1]
    }
}

fn main() {
    println!("{:#?}", Solution::unique_paths_with_obstacles(vec![vec![0,0,0],vec![0,1,0],vec![0,0,0]]));   // 2
}

File: 63/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(uniquePathsWithObstacles([][]int{{0,1},{0,0}}))	// 1
	fmt.Println(uniquePathsWithObstacles([][]int{{0,0,0},{0,1,0},{0,0,0}})) // 2
}

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	m, n := len(obstacleGrid), len(obstacleGrid[0])
	if obstacleGrid[0][0] == 1 || obstacleGrid[m-1][n-1] == 1{
		return 0
	}
	dp := make([][]int, m)
	for i := range dp {
		dp[i] = make([]int, n)
	}

	dp[0][0] = 1
	// 第一列
	for i := 1; i < m; i++ {
		if obstacleGrid[i][0] == 0 {
			dp[i][0] = dp[i-1][0]
		}
	}
	// 第一行
	for i := 1; i < n; i++ {
		if obstacleGrid[0][i] == 0 {
			dp[0][i] = dp[0][i-1]
		}
	}
	// i j 等于左和上的和
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if obstacleGrid[i][j] == 0 {
				dp[i][j] = dp[i-1][j] + dp[i][j-1]
			}
		}
	}

	return dp[m-1][n-1]
}